language/move-prover/bytecode/src/mutation_tester.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Transformation which mutates code for mutation testing in various ways
+//!
+//! This transformation should run after code is translated to bytecode, but before any
+//!  other bytecode modification
+//! It emits instructions in bytecode format, but with changes made
+//! Note that this mutation does nothing if mutation flags are not enabled
+
+use crate::{
+    function_data_builder::FunctionDataBuilder,
+    function_target::FunctionData,
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{Bytecode, Operation},
+    options::ProverOptions,
+};
+
+use move_model::{exp_generator::ExpGenerator, model::FunctionEnv, model::GlobalEnv};
+
+pub struct MutationTester {}
+
+struct MutationCounter{ 
+    value : usize
+}
+
+impl MutationTester {
+    pub fn new() -> Box<Self> {
+        Box::new(Self {})
+    }
+}
+
+impl FunctionTargetProcessor for MutationTester {
+    fn initialize(&self, global_env: &GlobalEnv, _targets: &mut FunctionTargetsHolder) {
+        let options = ProverOptions::get(global_env);
+        global_env.set_extension(MutationCounter{value : options.mutas});
+    }
+
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv<'_>,
+        data: FunctionData,
+    ) -> FunctionData {
+        use Bytecode::*;
+
+        if fun_env.is_native() {
+            // Nothing to do
+            return data;
+        }
+
+        let mut builder = FunctionDataBuilder::new(fun_env, data);
+        let code = std::mem::take(&mut builder.data.code);
+
+        // Emit trace instructions for parameters at entry.
+        builder.set_loc(builder.fun_env.get_loc().at_start());
+
+        for bc in code {
+            match &bc {
+                Call(attrid, indices, Operation::Add, tempindices, ab) => {
+                    let global_env = fun_env.module_env.env;
+                    let mc = global_env.get_extension::<MutationCounter>().unwrap().value;
+                    if mc == 1 {
+                        builder.emit(Call((*attrid).clone(), (*indices).clone(), 
+                            Operation::Sub, (*tempindices).clone(), (*ab).clone()));
+                    } else {
+                        builder.emit(bc.clone());
+                    }
+                    if mc > 0 {
+                        global_env.set_extension(MutationCounter{value : mc-1});
```

Comment:
I am not sure why a `set_extension/get_extension` is needed here. All you need seems to be a mutable usize here so why not just declare `let mut counter = options.mutas;` in the `process` function and  use `counter -= 1;` within the loop?

Reply:
Because I need this to be "global" across all processes.  Having a local mutable variable would not be kept across different functions (I tested it to see)

Reply:
First I don't quite see why this needs to be "global" across all processes. Second, wouldn't this code leads to a race condition? At least you should guarded it with a mutex or some kind of locking if you really want it to be shared across processes.

Reply:
This needs to be global since I want to mutate exactly one add operation instance in the given program (at a time, if you will), which, in my testing experience, means I need to have a counter that's global across operations / processes.  My rough understanding is that each function gets one process, so I need to make sure that each time I see an add operation across _all_ functions (aka processes in this case iiuc).  As for the mutex, my understanding is that the global_env would manage that (which works under testing), though I could absolutely be wrong on that understanding

Reply:
Depending on how you invoke these multiple processes, you may actually having either
1) one `GlobalEnv` shared with multiple **threads**, or
2) one `GlobalEnv` created in each **process**.
(see the difference between multi-processing vs multi-threading)

If 1), I think there is a serious problem, because `GlobalEnv` is not designed for multi-threading. An evidence is:
```rust
struct GlobalEnv {
    extensions: RefCell<BTreeMap<TypeId, Box<dyn Any>>>,
}
```
Note that this is a `RefCell`, which means it is not designed for sharing across threads (you would expect a `Mutex` or `Arc` for multi-threading).

Reply:
Ah, ok, I know where the confusion is: we have different understanding about what a "process" is.

You want the counter to be applied to **all** functions, i.e., if you already mutated one function, you will not mutate the other. In this sense, the code is correct, but just a bit cumbersome.

Reply:
Yes, exactly (sorry for the slow reply, got distracted with bugfixing).  I must have misrepresented my understanding of a process and where this was being applied, sorry about that.  Do you have any recommendations on how to make the code less cumbersome?  I couldn't think / don't know of another way to approach the problem

Reply:
I would grab the `MutationCounter` before the `for`-loop and put the updates back after the `for`-loop.

Reply:
Hmm, would this work though?  My understanding is that if there are multiple instances of things I would want to mutate in the same function, I would only want to mutate one: and I thought that the process was over a single function (which may have, for instance, multiple add operations)

Reply:
The move-prover code is 100% not thread-safe, so using GlobalEnv extensions for storing state fits into the existing reality. We need a serious refactor of the transformation traits to support some kind if "memory" across transformations. This came up before, but the right solution is not clear. Storing in the env for now is the only way I know.